### PR TITLE
Add an extra bounding box padding around a sub-app

### PR DESCRIPTION
**Pull request by andrsd**
This allows adding extra bounding box around 1D meshes so that layered transfers can be done with 2D and 3D meshes.

### CODE REVIEW

**Comment by andrsd** [comment]
@permcody This is a suggested solution for #10966. I do not have a test yet, but I am gonna do one today.

**Review by friedmud — CHANGES_REQUESTED** [review]
I think this is backward.... this stuff is supposed to be in the _Transfer_.  Different Transfers will need different amounts of inflation / etc.

For instance: note that Transfers already have `displaced_source_mesh` and `displaced_target_mesh` parameters... having one here doesn't make any sense (it's not clear if you're forcing the sub-apps to _compute_ on the displaced mesh or what you're doing).

**Comment by andrsd** [comment]
> it's not clear if you're forcing the sub-apps to compute on the displaced mesh or what you're doing

I think that `use_displaced_mesh = true` ([see the test file](https://github.com/idaholab/moose/pull/10967/commits/738fc0687ff519885b72e7497bc362efcc87e565#diff-65f18010e24decb9866a505f850c9779R92)) is pretty self-explanatory.

The problem is that the layered transfer does:
```
if (app_box.contains_point(*node))
```
where `app_box` for 1D meshes is something like (0, 0, 1)-(0, 0, 2) and good luck hitting that.

Moreover, `MultiApp` already has "bounding_box_inflation" parameter, so it made sense to be consistent and add the parameter there.

**Comment by friedmud** [comment]
I still disagree about the displaced mesh.

You could have one Transfer where you want to use the displaced mesh and another where you don't.  It doesn't make sense to set it at the MultiApp level.

An app's "bounding box inflation" makes sense to me.  I think that parameter is currently correct.  I could also see that a Transfer might want some _extra_ inflation or something.

**Comment by andrsd** [comment]
So the `use_displaced_mesh` parameter was not originally doing anything. The reason I made that do something was this line (in the transfer):
```
BoundingBox app_box = _multi_app->getBoundingBox(i);
```
This would give me the bounding box of an undisplaced mesh only, which was incorrect in my case.  What do you think about passing a parameter into this method that would determine if bounding box we are requesting should come from the displaced or undisplaced mesh? I agree that `use_displaced_mesh` parameter is a hack.

**Comment by friedmud** [comment]
Yep - sounds good

**Comment by permcody** [comment]
Well you are assigned to his PR. You'll need to go in an review and merge
then.

**Comment by andrsd** [comment]
1. rebased on current devel
2. fixed the hack so that `MultiApp::getBoundingBox` takes and extra parameter
3. found some brain damage in `MultiAppTransfer` parameters, so fixed that too (see d73ad60).

File: framework/src/multiapps/MultiApp.C [hunk 1]
```diff
@@ -360,21 +366,27 @@ MultiApp::restore()
...
+  FEProblemBase & fe_problem_base = _apps[local_app]->getExecutioner()->feProblem();
+  MooseMesh & mesh = (displaced_mesh && fe_problem_base.getDisplacedProblem().get() != NULL)
+                         ? fe_problem_base.getDisplacedProblem()->mesh()
+                         : fe_problem_base.mesh();
 
   BoundingBox bbox = {};
   {
     Moose::ScopedCommSwapper swapper(_my_comm);
-    bbox = MeshTools::create_bounding_box(problem.mesh());
+    bbox = MeshTools::create_bounding_box(mesh);
```
Comment (friedmud) [inline framework/src/multiapps/MultiApp.C]: One issue here: you are doing this every time for each Transfer.  I wonder how long this takes... I think it's a complete iteration over all the nodes in the mesh....

Comment (friedmud) [reply framework/src/multiapps/MultiApp.C]: We usually store a bounding box on MooseMesh... can't we reuse that here?  (Maybe just get that one and inflate it slightly?)

**Review by friedmud — COMMENTED** [review]
Some questions on efficiency...

File: framework/src/multiapps/MultiApp.C (hunk 1 above)
Comment (andrsd) [reply framework/src/multiapps/MultiApp.C]: If the mesh moved, we have to get the up-to-date bounding box. That's my understanding of this. 

The only bounding box on `MooseMesh` I see is the one returned by `MooseMesh::getInflatedProcessorBoundingBox`. Am I looking at the wrong place?  Also, this one should give back the global one, correct?

I can throw a patch on top of this to re-compute the BBox only if `displaced_mesh` is true and pre-cache the BBox the first time we call this.

Comment (friedmud) [reply framework/src/multiapps/MultiApp.C]: Ah - you're right.

In that case - call `getInflatedBoundingBox()`... and cache the result when you can.

**Review by friedmud — APPROVED** [review]

**Pull request merged**